geopandas/geodataframe.py

```diff
@@ -1433,17 +1444,45 @@ def __getitem__(self, key):
                 result.__class__ = DataFrame
         return result
 
+    def _persist_old_default_geometry_colname(self):
+        """Internal util to temporarily persist the default geometry column
+        name of 'geometry' for backwards compatibility."""
+        # self.columns check required to avoid this warning in __init__
+        if self._geometry_column_name is None and "geometry" not in self.columns:
```

Comment:
I think this is not only needed in the init, but for example also in:

```
df = geopandas.read_file(geopandas.datasets.get_path("naturalearth_lowres"))
df["geom2"] = df.geometry.centroid
df = df.set_geometry("geom2")
subset = df[["pop_est", "geometry"]].copy()
assert subset._geometry_column_name is None
subset["geometry"] = df.geom2.values
```

Although on second thought, should we actually warn here? Does that currently set the default geometry column? (I am running this on the PR's branch)

Reply:
So currently this snippet doesn't set the geometry column name on `main` and it also doesn't on this branch - but only because `geometry` already exists in the columns. If I modify this slightly, this PR warns and changes the geometry column:
```python
df = geopandas.read_file(geopandas.datasets.get_path("naturalearth_lowres"))
df["geom2"] = df.geometry.centroid
df["geom3"] = df['geometry']
df = df.set_geometry("geom2")
subset = df[["pop_est", "geom3"]].copy()
assert subset._geometry_column_name is None
subset["geometry"] = df.geom2.values
assert subset._geometry_column_name =="geometry" # this is bad
```
This is kind of edge casing I was worried about trying to issue a deprecation warning on this